experimental/rsocket-src/transports/TcpConnectionFactory.cpp

```diff
@@ -0,0 +1,128 @@
+      std::unique_ptr<ScopedEventBaseThread> eventBaseThread)
+      : fd_(std::move(fd)), eventBaseThread_(std::move(eventBaseThread)) {}
+
+  ~FramedDuplexConnectionOnThread() {
+    LOG(INFO) << "FramedDuplexConnectionOnThread => destroy";
+  }
+
+  void setInput(
+      std::shared_ptr<Subscriber<std::unique_ptr<folly::IOBuf>>> framesSink) {
+    fd_->setInput(std::move(framesSink));
+  }
+
+  std::shared_ptr<Subscriber<std::unique_ptr<folly::IOBuf>>> getOutput() {
+    return fd_->getOutput();
+  }
+
+ private:
+  std::unique_ptr<FramedDuplexConnection> fd_;
+  std::unique_ptr<ScopedEventBaseThread> eventBaseThread_;
+};
+
+// create new ScopedEventBaseThread
+// create new EventBase from it
+// create new AsyncSocket
+// connect
+// create FramedDuplexConnection
+// TODO create variant that takes an existing EventBase
+class SocketConnectorAndCallback : public AsyncSocket::ConnectCallback {
+ public:
+  SocketConnectorAndCallback(OnConnect onConnect, SocketAddress addr)
+      : addr_(addr),
+        onConnect_(onConnect),
+        eventBaseThread_(std::make_unique<ScopedEventBaseThread>()) {}
+
+  ~SocketConnectorAndCallback() {
+    LOG(INFO) << "SocketConnectorAndCallback => destroy";
+  }
+
+  void connect() {
+    // now start the connection asynchronously
+    eventBase_->runInEventBaseThreadAndWait([this]() {
+      LOG(INFO) << "ConnectionFactory => starting socket";
+      socket_.reset(new folly::AsyncSocket(eventBase_));
+
+      LOG(INFO) << "ConnectionFactory => attempting connection to "
+                << addr_.describe() << std::endl;
+
+      socket_->connect(this, addr_);
+
+      LOG(INFO) << "ConnectionFactory  => DONE connect";
+    });
+  }
+
+ private:
+  SocketAddress addr_;
+  folly::AsyncSocket::UniquePtr socket_;
+  OnConnect onConnect_;
+  std::unique_ptr<ScopedEventBaseThread> eventBaseThread_;
+  EventBase* eventBase_{eventBaseThread_->getEventBase()};
+
+  void connectSuccess() noexcept {
+    LOG(INFO) << "ConnectionFactory => socketCallback => Success";
+
+    // safe way to call 'delete this'
+    auto uThis = std::unique_ptr<SocketConnectorAndCallback>(this);
+
+    auto connection = std::make_unique<TcpDuplexConnection>(
+        std::move(socket_), inlineExecutor(), Stats::noop());
+    auto framedConnection = std::make_unique<FramedDuplexConnection>(
+        std::move(connection), inlineExecutor());
+
+    auto framedConnectionOnThread =
+        std::make_unique<FramedDuplexConnectionOnThread>(
+            std::move(framedConnection), std::move(eventBaseThread_));
+
+    // callback with the connection now that we have it
+    onConnect_(std::move(framedConnectionOnThread), *eventBase_);
+  }
+
+  void connectErr(const AsyncSocketException& ex) noexcept {
+    LOG(INFO) << "ConnectionFactory => socketCallback => ERROR => " << ex.what()
+              << " " << ex.getType() << std::endl;
+
+    delete this;
+  }
+};
+
+TcpConnectionFactory::TcpConnectionFactory(std::string host, uint16_t port)
```

Comment:
you can inline the bodies of the methods in the cpp file.
the only separation of declarations and bodies is done in header files.